clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -610,8 +610,12 @@ public void unmute(String id) {
     }
 
     private void unmute(KafkaChannel channel) {
-        explicitlyMutedChannels.remove(channel);
         channel.unmute();
+        // For SocketServer, the above call may not unmute depending on the mute state. Remove the channel from
+        // explicitlyMutedChannels only if the channel has been actually unmuted.
+        if (!channel.isMute()) {
```

Comment:
Perhaps it would be better if `KafkaChannel.unmute` returned a boolean. Then you could do `if (channel.unmute())` instead of two calls. You could also change the method name to `maybeUnmute` to make it more obvious.

Reply:
Done